internal/handlers/common/route.go

```diff
@@ -0,0 +1,156 @@
+//  * return normal response body;
+//  * return protocol error (*common.Error, possibly wrapped) - it will be returned to the client;
+//  * return any other error - it will be returned to the client as InternalError before terminating connection.
+//
+// They should not panic on bad input, but may do so in "impossible" cases.
+// They also should not use recover(). That allows us to use fuzzing.
+// (Panics terminate the connection without a response on a different level.)
+func Route(h Handler, ctx context.Context, reqHeader *wire.MsgHeader, reqBody wire.MsgBody,
+) (resHeader *wire.MsgHeader, resBody wire.MsgBody, closeConn bool, result *string) {
+	var command string
+	defer func() {
+		if result == nil {
+			result = pointer.ToString("panic")
+		}
+	}()
+
+	resHeader = new(wire.MsgHeader)
+	var err error
+	switch reqHeader.OpCode {
+	case wire.OP_MSG:
+		msg := reqBody.(*wire.OpMsg)
+		var document *types.Document
+		document, err = msg.Document()
+		command = document.Command()
+
+		if err == nil {
+			resHeader.OpCode = wire.OP_MSG
+			resBody, err = handleOpMsg(h, ctx, msg, command)
+		}
+
+	case wire.OP_QUERY:
+		query := reqBody.(*wire.OpQuery)
+		resHeader.OpCode = wire.OP_REPLY
+		resBody, err = h.CmdQuery(ctx, query)
+
+	case wire.OP_REPLY:
+		fallthrough
+	case wire.OP_UPDATE:
+		fallthrough
+	case wire.OP_INSERT:
+		fallthrough
+	case wire.OP_GET_BY_OID:
+		fallthrough
+	case wire.OP_GET_MORE:
+		fallthrough
+	case wire.OP_DELETE:
+		fallthrough
+	case wire.OP_KILL_CURSORS:
+		fallthrough
+	case wire.OP_COMPRESSED:
+		fallthrough
+	default:
+		err = lazyerrors.Errorf("unexpected OpCode %s", reqHeader.OpCode)
+	}
+
+	// set body for error
+	if err != nil {
+		switch resHeader.OpCode {
+		case wire.OP_MSG:
+			protoErr, recoverable := ProtocolError(err)
+			closeConn = !recoverable
+			var res wire.OpMsg
+			err = res.SetSections(wire.OpMsgSection{
+				Documents: []*types.Document{protoErr.Document()},
+			})
+			if err != nil {
+				panic(err)
+			}
+			resBody = &res
+			result = pointer.ToString(protoErr.Code().String())
+
+		case wire.OP_QUERY:
+			fallthrough
+		case wire.OP_REPLY:
+			fallthrough
+		case wire.OP_UPDATE:
+			fallthrough
+		case wire.OP_INSERT:
+			fallthrough
+		case wire.OP_GET_BY_OID:
+			fallthrough
+		case wire.OP_GET_MORE:
+			fallthrough
+		case wire.OP_DELETE:
+			fallthrough
+		case wire.OP_KILL_CURSORS:
+			fallthrough
+		case wire.OP_COMPRESSED:
+			fallthrough
+		default:
+			// do not panic to make fuzzing easier
+			closeConn = true
+			result = pointer.ToString("unexpected")
+			return
+		}
+	}
+
+	// TODO Don't call MarshalBinary there. Fix header in the caller?
+	// https://github.com/FerretDB/FerretDB/issues/273
+	b, err := resBody.MarshalBinary()
+	if err != nil {
+		result = nil
+		panic(err)
+	}
+	resHeader.MessageLength = int32(wire.MsgHeaderLen + len(b))
+
+	// resHeader.RequestID = atomic.AddInt32(&h.lastRequestID, 1) // XXX: not set anywhere
```

Comment:
We should set it though

Reply:
Let's do this PR refactoring only.
Need to know how to generate it, who else should have access to it, in which logs to pass, probably let logger be in the context etc.
WDYT?

Reply:
Sorry, I don’t follow. That field is being set now: https://github.com/FerretDB/FerretDB/blob/6f00c8dc8059c1d1707bdf5c23d61a19e29c19d7/internal/handlers/pg/handler.go#L192 I don’t see what prevents us from uncommenting this line back to keep it as it was before

Reply:
then the handler will have an interface LastRequestID(), it's ok?

Reply:
moved to `clentconn` as route now is a `clientconn`-s.